lib/scanner/list-scanner/list-scanner.c

```diff
@@ -0,0 +1,202 @@
+ *
+ */
+#include "scanner/list-scanner/list-scanner.h"
+#include "str-repr/encode.h"
+#include "str-repr/decode.h"
+#include "str-utils.h"
+
+#include <string.h>
+#include <stdarg.h>
+
+static void
+_free_input(ListScanner *self)
+{
+  if (self->argv)
+    {
+      if (self->free_argv_payload)
+        g_ptr_array_foreach(self->argv_buffer, (GFunc) g_free, NULL);
+    }
+  g_ptr_array_set_size(self->argv_buffer, 0);
+}
+
+static void
+_store_input(ListScanner *self, gint argc, gchar **argv, gboolean free_argv)
+{
+  self->argc = argc;
+  self->argv = (gchar **) self->argv_buffer->pdata;
+  self->free_argv_payload = free_argv;
+  self->current_arg_ndx = 0;
+  self->current_arg = self->argv[self->current_arg_ndx];
+}
+
+void
+list_scanner_input_va(ListScanner *self, const gchar *arg1, ...)
+{
+  va_list va;
+  gint i;
+  const char *arg;
+
+  _free_input(self);
+
+  va_start(va, arg1);
+  for (i = 0, arg = arg1; arg; i++, arg = va_arg(va, const gchar *))
+    {
+      g_ptr_array_add(self->argv_buffer, g_strdup(arg));
+    }
+  g_ptr_array_add(self->argv_buffer, NULL);
+  va_end(va);
+  _store_input(self, i, (gchar **) self->argv_buffer->pdata, TRUE);
+}
+
+void
+list_scanner_input_gstring_array(ListScanner *self, gint argc, GString *argv[])
+{
+  gint i;
+
+  for (i = 0; i < argc; i++)
+    {
+      g_ptr_array_add(self->argv_buffer, argv[i]->str);
+    }
+  g_ptr_array_add(self->argv_buffer, NULL);
+  _store_input(self, argc, (gchar **) self->argv_buffer->pdata, FALSE);
+}
+
+static gboolean
+_is_finished_with_args(ListScanner *self)
+{
+  return self->current_arg_ndx >= self->argc;
+}
+
+static gboolean
+_parse_unquoted_empty_value(ListScanner *self)
+{
+  if (*self->current_arg == ',')
+    {
+      self->current_arg = self->current_arg + 1;
+      return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+_parse_value_from_current_position(ListScanner *self)
+{
+  StrReprDecodeOptions options = {
+    0,
+    .delimiter_chars = ",",
+  };
+  const gchar *end;
+
+  /* NOTE: we always try to parse elements, even if the string is malformed,
+   * otherwise we would be losing data.  Prefer to have data in an
+   * incorrectly formatted way, than no data at all. */
+  if (!str_repr_decode_with_options(self->value, self->current_arg, &end, &options))
+    {
+      g_string_truncate(self->value, 0);
+      g_string_append_len(self->value, self->current_arg, end - self->current_arg);
+    }
+  self->current_arg = end;
+
+  return TRUE;
+}
+
+static gboolean
+_locate_next_non_empty_arg(ListScanner *self)
+{
+  while (!_is_finished_with_args(self) && self->current_arg[0] == 0)
+    {
+      self->current_arg_ndx++;
+      self->current_arg = self->argv[self->current_arg_ndx];
+    }
+  return !_is_finished_with_args(self);
+}
+
+gboolean
+list_scanner_scan_next(ListScanner *self)
+{
+  g_string_truncate(self->value, 0);
+
+  while (_locate_next_non_empty_arg(self))
+    {
+      if (_parse_unquoted_empty_value(self))
+        {
+          ;
```

Comment:
I think adding this empty statement is superfluous, the empty block statement is enough.
(not a blocking issue, just a note)